{% extends "base.html" %}
{% block content %}
<!-- Page Heading -->
<div class="d-sm-flex align-items-center justify-content-between mb-4">
    <h1 class="h3 mb-0 text-gray-800">Classification</h1>
</div>
<p class="mb-4">
    The following interface serves as a visualization of the results produced by TSS-CAPTUR. The possibility exists between an <a href="overview.html">overview</a>> of all results or the detailed results of the <a href="classified.html">classification</a> or the <a href="terminators.html">terminator allocation</a> steps of TSS-CAPTUR. The possibility of some TSS being ignored by TSS-CAPTUR due to their short lenght exists. They are found <a href="avoided.html">here</a> . </p>
    <p class="mb-4">The current page presents a detailed view on the results on the transcript classification. TSS-CAPTUR uses two programs (CNIT and QRNA) to classify the transcripts in either coding (COD) or non-coding (RNA) transcripts. 
        The following table allows the access to the final classification and some intermediate steps before it, such as which organism was used for the pairwiese alignment for QRNA. 
        Since both programs used different scoring functions, a normalized z-score is given. If both programs agree on the predicion, the subcoordinates of the program with the highest z-score are used. 
        If the predictions differ, TSS-CAPTUR labels the transcipt as a coding sequence. </p>
<!-- Content Row -->
  <div class="card shadow mb-4">
    <div class="card-header py-3">
        <h6 class="m-0 font-weight-bold text-primary">Detailed view on the classification</h6>
    </div>
    <div class="card-body">
        <div class="table-responsive">
            <table class="table table-bordered" id="dataTableClass" width="100%" cellspacing="0">
                
            </table>
        </div>
    </div>
</div>


<script type='text/javascript'>
    index = localStorage.getItem("genome") ? localStorage.getItem("genome") : 0;
    genome_selected =genomes[index];
    let dataClass = classified[genome_selected];

    $(document).ready(function () {
        $("#dataTableClass").DataTable({
            "scrollX": true,
            "scrollY":$(window).height()*0.8,
            "scrollCollapse": true,
            columnDefs: [{ type: "natural", targets: 0 },  
            { render: function ( data, type, row ) {
        return typeof(data) == "number" ? data.toFixed(3) : data ;
            },
    targets: [ 6,7 ] }],
            order: [[0, "asc"]],
            data: dataClass,
            columns: [
            { title: "Transcript ID" },
            { title: "Position" },
            { title: "Strand" },
            { title: "Aligned to" },
            { title: "Same Prediction?" },
            { title: "Source" },
            { title: "Score of program" },
            { title: "Normalized score" },
            { title: "Final Class" },
            { title: "Start of Transcript" },
            { title: "Predicted Gene Start" },
            { title: "End of Transcript " },
            { title: "Predicted Gene End" },
            ],
        });
    });
    
    
</script>
{% endblock %}
