{% extends "base.html" %}
{% block content %}
<!-- Page Heading -->
<div class="d-sm-flex align-items-center justify-content-between mb-4">
    <h1 class="h3 mb-0 text-gray-800">Terminator prediction</h1>
</div>
<p class="mb-4">
    The following interface serves as a visualization of the results produced by TSS-CAPTUR. The possibility exists between an <a href="overview.html">overview</a>> of all results or the detailed results of the <a href="classified.html">classification</a> or the <a href="terminators.html">terminator allocation</a> steps of TSS-CAPTUR. The possibility of some TSS being ignored by TSS-CAPTUR due to their short lenght exists. They are found <a href="avoided.html">here</a> . </p>
    <p class="mb-4">The current page presents a detailed view on the results of the terminator allocation for RNA transcripts. TSS-CAPTUR predicts intrinsic as well as rho-dependent terminators using TransTermHP and RhoTermPredict, respectively. 
        If a terminator is found within a distance of 600 nt to the identified TSS, the terminator is allocated to the corresponding transcript. The terminator is evaluated wrt. to the predicted coordinates of the classification to ensure a terminator close to the 3' end of the predicted gene. 
    If many terminators are allocated to the same TSS, only the best-scored terminator is chosen. If this terminator is rho-dependent, an intrinsic terminator is search within a window of 150 nts. 
    If a terminator is allocated to many TSSs, the terminators are ranked wrt. to the TSS and the computed score. The 3'-end coordinates of only those transcripts with the best-ranked terminators are updated. 
</p>
<!-- Content Row -->
<div class="card shadow mb-4">
    <div class="card-header py-3">
        <h6 class="m-0 font-weight-bold text-primary">Detailed view of terminator prediction</h6>
    </div>
    <div class="card-body">
        <div class="table-responsive">
            <table class="table table-bordered" id="dataTableTerms" width="100%" cellspacing="0">
                
            </table>
        </div>
    </div>
</div>


<script type='text/javascript'>

    index = localStorage.getItem("genome") ? localStorage.getItem("genome") : 0;
    genome_selected =genomes[index];
    let dataTerms = terminators[genome_selected];
    $(document).ready(function () {
        $(document).ready(function () {
            $("#dataTableTerms").DataTable({
                "scrollX": true,
                "scrollY":$(window).height()*0.8,
                "scrollCollapse": true,
                columnDefs: [{ type: "natural", targets: [0,7] },{ render: function ( data, type, row ) {
        return parseFloat(data).toFixed(3) ;
            },
    targets: [ 7 ] }],
                order: [[0, "asc"]],
                data: dataTerms,
                columns: [
                { title: "Transcript ID" },
                { title: "Strand" },
                { title: "Class" },
                { title: "Gene Start" },
                { title: "Gene End" },
                { title: "Type of Terminator" },
                { title: "Terminator ID" },
                { title: "Score" },
                { title: "Start Term" },
                { title: "End Term" },
                { title: "Rank" },
                { title: "Additional Terminator" },
                { title: "Start with Terminator" },
                { title: "End with Terminator" },
                ],
            });
        });
    });

    

    
</script>

{% endblock %}
